### PR TITLE
Delete idle reservations

### PR DESCRIPTION
**Pull request by bhroam**
#### Describe Bug or Feature
Resources are held by reservations whether they are used or not.  If all jobs end in a running reservations, the resources sit idle.  This feature adds the ability for a reservation to end after it sits idle for a certain period of time.

#### Describe Your Change
I added the reservation attribute delete_idle_time to hold the amount of time for a reservation to sit idle before ending.  When a reservation starts or a job in the reservation ends, we evaluate if there are any jobs in it.  If there aren't, we add a task to the timed task list to end the reservation.  When the task fires, we check to see if the reservation is still empty, and if it is we end it.

I also fixed an issue where a standing degraded reservation was not going back into the confirmed state if all the nodes are up in the current occurrence.

#### Link to Design Doc
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1479934178/Delete+empty+running+reservations+automatically

#### Attach Test and Valgrind Logs/Output
[resv.log](https://github.com/PBSPro/pbspro/files/4156446/resv.log)

### CODE REVIEW

File: src/lib/Libattr/master_resv_attr_def.xml [hunk 1]
```diff
@@ -780,6 +780,23 @@
...
+	<member_at_encode>encode_time</member_at_encode>
+	<member_at_set>set_l</member_at_set>
+	<member_at_comp>comp_l</member_at_comp>
+	<member_at_free>free_null</member_at_free>
+	<member_at_action>NULL_FUNC</member_at_action>
+	<member_at_flags><both>READ_WRITE</both></member_at_flags>
+	<member_at_type><both>ATR_TYPE_LONG</both></member_at_type>
+	<member_at_parent>PARENT_TYPE_RESV</member_at_parent>
+	<member_verify_function>
+	<ECL>NULL_VERIFY_DATATYPE_FUNC</ECL>
```
Comment (vstumpf) [inline src/lib/Libattr/master_resv_attr_def.xml]: Should there be a verify function? Maybe verify_datatype_long?

File: src/server/req_quejob.c [hunk 2]
```diff
@@ -2667,6 +2667,12 @@ req_resvSub(struct batch_request *preq)
 	presv->ri_wattr[(int)RESV_ATR_mtime].at_flags |= ATR_VFLAG_SET |
 		ATR_VFLAG_MODIFY | ATR_VFLAG_MODCACHE;
 
+	if (presv->ri_wattr[(int) RESV_ATR_convert].at_flags & ATR_VFLAG_SET)
+		if (!(presv->ri_wattr[(int) RESV_ATR_del_idle_time].at_flags & ATR_VFLAG_SET)) {
```
Comment (vstumpf) [inline src/server/req_quejob.c]: You can combine these two if statements

File: src/server/req_rescq.c [hunk 3]
```diff
@@ -799,22 +817,22 @@ req_confirmresv(struct batch_request *preq)
 	 * Notify all interested parties that the reservation
 	 * is moving from state UNCONFIRMED to CONFIRMED
 	 */
-	if (presv->ri_brp) {
-		presv = find_resv(presv->ri_qs.ri_resvID);
-		if (presv->ri_wattr[(int)RESV_ATR_convert].at_val.at_str != NULL) {
-			rc = cnvrt_qmove(presv);
-			if (rc != 0) {
-				snprintf(buf, sizeof(buf), "%.240s FAILED",  presv->ri_qs.ri_resvID);
+		if (presv->ri_brp) {
```
Comment (vstumpf) [inline src/server/req_rescq.c]: The indentation is off on this block.

File: test/tests/functional/pbs_reservations.py [hunk 4]
```diff
@@ -67,15 +67,19 @@ def submit_reservation(self, select, start, end, user, rrule=None,
 
             a.update({ATTR_resv_rrule: rrule, ATTR_resv_timezone: tzone})
 
+        if extra_attrs is not None:
```
Comment (vstumpf) [inline test/tests/functional/pbs_reservations.py]: Can change this to just `if extra_attrs`

File: test/tests/functional/pbs_reservations.py [hunk 5]
```diff
@@ -67,15 +67,19 @@ def submit_reservation(self, select, start, end, user, rrule=None,
...
 
         return self.server.submit(r)
 
-    def submit_asap_reservation(self, user, jid):
+    def submit_asap_reservation(self, user, jid, extra_attrs=None):
         """
         Helper method to submit an ASAP reservation
         """
         a = {ATTR_convert: jid}
+        if extra_attrs is not None:
```
Comment (vstumpf) [inline test/tests/functional/pbs_reservations.py]: Same as above

File: src/lib/Libattr/master_resv_attr_def.xml (hunk 1 above)
Comment (bhroam) [reply src/lib/Libattr/master_resv_attr_def.xml]: We don't need one.  It should only be set by the server.  The verify functions are client side.

File: src/lib/Libutil/work_task.c [hunk 6]
```diff
@@ -180,47 +180,26 @@ delete_task_by_parm1_func(void *parm1, void (*func)(struct work_task *), enum wt
...
-		if (((ptask->wt_parm1 != NULL) && (ptask->wt_parm1 == parm1)) ||
-		    ((ptask->wt_func != NULL) && (ptask->wt_func == func))) {
-			delete_task(ptask);
-			if (option == DELETE_ONE)
-				return;
-		}
-		ptask = ptask_next;
-	}
+	for (i = 0; i < 3; i++) {
+		for (ptask = (struct work_task *) GET_NEXT(task_lists[i]); ptask; ptask = ptask_next) {
```
Comment (vstumpf) [inline src/lib/Libutil/work_task.c]: If you want to, you can rewrite this as 
```suggestion
		for (ptask = (struct work_task *) GET_NEXT(task_lists[i]); ptask; ptask = (struct work_task *) GET_NEXT(ptask->wt_linkall)) {
```
And you can get rid of ptask_next

**Comment by bhroam** [comment]
I made some changes to delete_task_by_parm1_func() to clean it up (and fix it).  Time4occurrenceFinish() had an issue if the standing reservation so much time in the deleting job state that the reservation started late.  It would actually delete that occurrence.  Lastly I needed to change the ECL validation function from long to time since you can use the HH:MM:SS form instead of SS.

**Pull request merged**